vllm/v1/core/scheduler.py

```diff
@@ -58,8 +60,12 @@ def __init__(
 
         # req_id -> Request
         self.requests: Dict[str, Request] = {}
-        # Priority queues for requests.
-        self.waiting: Deque[Request] = deque()
+        # NOTE: Priority queues for requests.
+        # With list, we can safely pop the index
+        # of a request that are yet to be ready (in this case,
+        # the one that uses guided decoding) while still maintaining
+        # the order of all requests in existing waiting queue.
```

Comment:
Not quite -- we pop a ready request from the middle, even when higher priority guided decoding requests remain in the list ahead of it since they're not ready.  we don't pop the not-ready ones